numpy/core/src/umath/matmul.c.src

```diff
@@ -0,0 +1,393 @@
+            ip2 -= ib2_n;
+            op  +=  os_p;
+            ip2 += is2_p;
+        }
+        op -= ob_p;
+        ip2 -= ib2_p;
+        ip1 += is1_m;
+        op  +=  os_m;
+    }
+}
+
+/**end repeat**/
+
+/**begin repeat
+ *  #TYPE = FLOAT, DOUBLE, LONGDOUBLE, HALF,
+ *          CFLOAT, CDOUBLE, CLONGDOUBLE,
+ *          UBYTE, USHORT, UINT, ULONG, ULONGLONG,
+ *          BYTE, SHORT, INT, LONG, LONGLONG,
+ *          BOOL#
+ *  #typ = npy_float,npy_double,npy_longdouble, npy_half,
+ *         npy_cfloat, npy_cdouble, npy_clongdouble,
+ *         npy_ubyte, npy_ushort, npy_uint, npy_ulong, npy_ulonglong,
+ *         npy_byte, npy_short, npy_int, npy_long, npy_longlong,
+ *         npy_bool#
+ * #IS_COMPLEX = 0, 0, 0, 2, 1, 1, 1, 0*11#
+ * #USEBLAS = 1, 1, 0, 0, 1, 1, 0*12#
+ */
+
+
+NPY_NO_EXPORT void
+@TYPE@_matmul(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED(func))
+{
+    npy_intp dOuter = *dimensions++;
+    npy_intp iOuter;
+    npy_intp s0 = *steps++;
+    npy_intp s1 = *steps++;
+    npy_intp s2 = *steps++;
+    npy_intp dm = dimensions[0];
+    npy_intp dn = dimensions[1];
+    npy_intp dp = dimensions[2];
+    npy_intp is1_m=steps[0], is1_n=steps[1], is2_n=steps[2], is2_p=steps[3],
+         os_m=steps[4], os_p=steps[5];
+#if @USEBLAS@ && defined(HAVE_CBLAS)
+    npy_intp sz = sizeof(@typ@);
+    npy_bool special_case = (dm == 1 || dn == 1 || dp == 1);
+    npy_bool scalar_out = (dm == 1 && dp == 1);
+    npy_bool scalar_vec = (dn == 1 && (dp == 1 || dm == 1));
+    npy_bool too_big_for_blas = (dm > NPY_MAX_INT || dn > NPY_MAX_INT ||
+                                 dp >= NPY_MAX_INT);
+    npy_bool i1blasable = (is1_m == sz && BLAS_ABLE(dm, is1_n, sz)) ||
+                          (is1_n == sz && BLAS_ABLE(dn, is1_m, sz));
+    npy_bool i2blasable = (is2_n == sz && BLAS_ABLE(dn, is2_p, sz)) ||
+                          (is2_p == sz && BLAS_ABLE(dp, is2_n, sz));
+    npy_bool o_c_blasable = (os_p == sz) && BLAS_ABLE(dp, os_m, sz);
+    npy_bool o_f_blasable = (os_m == sz) && BLAS_ABLE(dm, os_p, sz);
+    npy_bool vector_matrix = ((dm == 1) && i2blasable);
+    npy_bool matrix_vector = ((dp == 1)  && i1blasable);
+#endif
+
+    if (dn == 0) {
+        /* No operand, need to zero the output */
+        for (iOuter = 0; iOuter < dOuter; iOuter++,
+                        args[0] += s0, args[1] += s1, args[2] += s2) {
+            npy_intp m, p;
+            char *op=args[2];
+            for (m = 0; m < dm; m++) {
+                for (p = 0; p < dp; p++) {
+#if @IS_COMPLEX@ == 1
+                    (*(@typ@ *)op).real = 0;
+                    (*(@typ@ *)op).imag = 0;
+#else
+                    *(@typ@ *)op = 0;
+#endif
+                    op  +=  os_p;
+                }
+                op  +=  os_m - (os_p * dp);
+            }
+        }
+        return;
+    }
+    for (iOuter = 0; iOuter < dOuter; iOuter++,
+                         args[0] += s0, args[1] += s1, args[2] += s2) {
```

Comment:
You could save a lot of branching by transposing this loop with the if statements - but the compiler is probably smart enough to do this anyway.

Reply:
ok, will try this. It will be cleanest if I unify all the inner function signatures and make a `typedef func*`

Reply:
Sure you don't want to leave this to a follow-up PR?

Reply:
ok, will stop with last commit

Reply:
> Indirecting via a function pointer will lead to branching just like you have here

My intention was to do all the logic, then the loop so something like
```
if .... :
  func = f1
elif ... :
    func = f2
/* Done with logic, now enter the loop */
for ...
    func(args)
```
but I will leave this for now.

Anything else (other than the last rename) need fixing?